src/VCZ.cpp

```diff
@@ -0,0 +1,267 @@
+        array_;
+
+    void cache_first_chunk() {
+        const auto rank = store_.rank();
+        const auto shape = store_.domain().shape();
+        const auto chunk_shape = store_.chunk_layout().value().read_chunk().shape();
+        const std::vector<Index> chunk_starts(rank);
+        std::vector<Index> chunk_ends(rank);
+
+        for (Index dim_index = 0; dim_index < rank; dim_index++)
+            chunk_ends[dim_index] = std::min(shape[dim_index], chunk_shape[dim_index]);
+
+        array_ = tensorstore::Read(store_ | tensorstore::AllDims().HalfOpenInterval(chunk_starts, chunk_ends)).value();
+    }
+};
+
+class VczClass::Impl {
+   public:
+    Impl(std::string& t_vczFileName,
+         std::vector<std::string>& t_sampleInModel) {
+        m_model_sample_count = t_sampleInModel.size();
+        tensorstore::Spec input_spec;
+        input_spec = tensorstore::Spec::FromJson(
+                         {
+                             {"driver", "zarr"},
+                             {"kvstore",
+                              {{"driver", "file"},
+                               {"path", t_vczFileName + "/call_genotype"}}},
+                         })
+                         .value();
+        auto call_genotype_store =
+            tensorstore::Open<int8_t, tensorstore::dynamic_rank,
+                              tensorstore::ReadWriteMode::read>(
+                input_spec, tensorstore::OpenMode::open,
+                tensorstore::ReadWriteMode::read)
+                .value();
+        m_genotype_shape = call_genotype_store.domain().shape();
+        m_call_genotype_array =
+            ChunkCacher<int8_t>(std::move(call_genotype_store));
+
+        input_spec = tensorstore::Spec::FromJson(
+                         {
+                             {"driver", "zarr"},
+                             {"kvstore",
+                              {{"driver", "file"},
+                               {"path", t_vczFileName + "/contig_length"}}},
+                         })
+                         .value();
+        m_contig_length_array = ChunkCacher<int64_t>(input_spec);
+
+        input_spec = tensorstore::Spec::FromJson(
+                         {
+                             {"driver", "zarr"},
+                             {"kvstore",
+                              {{"driver", "file"},
+                               {"path", t_vczFileName + "/variant_contig"}}},
+                         })
+                         .value();
+        m_variant_contig_array = ChunkCacher<int8_t>(input_spec);
+
+        input_spec = tensorstore::Spec::FromJson(
+                         {
+                             {"driver", "zarr"},
+                             {"kvstore",
+                              {{"driver", "file"},
+                               {"path", t_vczFileName + "/variant_position"}}},
+                         })
+                         .value();
+        m_variant_position_array = ChunkCacher<int8_t>(input_spec);
```

Comment:
Does this mean variant_positions must be int8_t values? This is too small if so. int32_t is safe.

Reply:
I chose the element types to match data types in the Zarr arrays as they are stored on my computer. I will see if using larger element types works.

Reply:
We'd need int32 to work with any reasonable benchmark data

Reply:
Just want to note that while working on the performance testing, I found that the element type template parameter needs to match the Zarr array's data type. The **element type template parameter** is a [type template parameter](https://en.cppreference.com/w/cpp/language/template_parameters#Type_template_parameter) that TensorStore uses as the data type of the TensorStore store. See [here](https://github.com/google/tensorstore/blob/master/tensorstore/open.h#L91) for example. Essentially, the TensorStore code is parameterized around the element type, which can be specified at compile time by the user.

TensorStore has an error if I try to use `int32_t` to read an `<i1` Zarr array. The current implementation requires the user to update the code, specifically the arguments to the type template parameters, to match the data types of the respective Zarr arrays. I currently don't know how to access data through TensorStore without specifying the element type somewhere at compile time.